File: rcm-core/src/cmds/open_with.rs
//! `@open-with` — Open the Windows "Open With → Choose another app" dialog.
//!
//! Spawns PowerShell in the background to invoke the `openas` shell verb
//! via `Shell.Application` COM — the exact equivalent of right-click →
//! "Open with" → "Choose another app" in Windows Explorer.
//!
//! Uses `spawn` (fire-and-forget) rather than `output` to avoid blocking
//! the Rust async runtime and to prevent the dialog activation from
//! being misinterpreted as a new right-click event.

use super::SystemCmdResult;
use crate::types::CommandPayload;

pub fn run(cmd: &CommandPayload) -> SystemCmdResult {
    let path = cmd.args.first().map(|s| s.as_str()).unwrap_or("");

    if path.is_empty() {
        return SystemCmdResult {
            success: false,
            message: "@open-with requires a file path argument".into(),
        };
    }

    // Escape single quotes for PowerShell string interpolation
    let escaped = path.replace('\'', "''");

    let script = format!(
        "$f=gi -LiteralPath '{escaped}';\
         (New-Object -ComObject Shell.Application).Namespace($f.DirectoryName).ParseName($f.Name).InvokeVerb('openas')"
    );

    match std::process::Command::new("powershell")
        .args(["-NoProfile", "-Command", &script])
        .spawn()
    {
        Ok(_) => SystemCmdResult {
            success: true,
            message: format!("OpenWith dialog launched for: {path}"),
        },
        Err(e) => SystemCmdResult {
            success: false,
            message: format!("OpenWith failed to launch: {e}"),
        },
    }
}

File: rcm-kit/src/menus/open-with.ts
import { OPEN_WITH } from "../consts"
import { t } from "../i18n"
import type { MenuItem, InvokeProps } from "../types"

/**
 * Win11 "Open with" — opens the Windows "Open With → Choose another app"
 * dialog for the selected file via the native `@open-with` system command.
 *
 * Falls back to the current directory when no file is selected
 * (background right-click).
 */
export function openWith(): MenuItem {
  return {
    key: "open-with",
    label: t("open.with"),
    icon: "🔽",
    match: ({ files }) => files.length === 1,
    action: (props: InvokeProps) => {
      const target = props.files[0]
      const path = target ? target.path : props.cwd
      if (!path || props.files.length > 1) return
      return {
        cmd: OPEN_WITH,
        args: [path],
        window: "Hidden",
      }
    },
  }
}
